libr/debug/debug.c

```diff
@@ -26,40 +26,150 @@ R_API void r_debug_info_free (RDebugInfo *rdi) {
+
+	/* remove all sw breakpoints for now. we'll set them back in stage 2
+	 *
+	 * this is necessary because while stopped we don't want any breakpoints in
+	 * the code messing up our analysis.
+	 */
+	if (!r_bp_restore (dbg->bp, false)) // unset sw breakpoints
 		return false;
+
+	/* if we are recoiling, tell r_debug_step that we ignored a breakpoint
+	 * event */
+	if (!dbg->swstep && dbg->recoil_mode != R_DBG_RECOIL_NONE) {
+		dbg->reason.bp_addr = 0;
+		return true;
 	}
-	r_debug_reg_sync (dbg, R_REG_TYPE_GPR, false);
-	ri = r_reg_get (dbg->reg, dbg->reg->name[R_REG_NAME_PC], -1);
-	dbg->reason.bpi = NULL;
-	if (ri) {
-		ut64 addr = r_reg_get_value (dbg->reg, ri);
-		recoil = r_bp_recoil (dbg->bp, addr - dbg->bpsize);
-		//eprintf ("[R2] Breakpoint recoil at 0x%"PFMT64x" = %d\n", addr, recoil);
-		if (recoil < 1)
-			recoil = 0; // XXX Hack :D
-		if (recoil) {
-			dbg->in_recoil = true;
-			dbg->reason.type = R_DEBUG_REASON_BREAKPOINT;
-			dbg->reason.bpi = r_bp_get_at (dbg->bp, addr-recoil);
-			dbg->reason.addr = addr - recoil;
-			r_reg_set_value (dbg->reg, ri, addr-recoil);
-			if (r_reg_get_value (dbg->reg, ri) != (addr-recoil)) {
-				eprintf ("r_debug_recoil: Cannot set program counter\n");
+
+	/* see if we really have a breakpoint here... */
+	b = r_bp_get_at (dbg->bp, pc - dbg->bpsize);
+	if (!b) /* we don't. nothing left to do */
+		return true;
+
+	/* set the pc value back */
+	pc -= b->size;
+	if (!r_reg_set_value (dbg->reg, pc_ri, pc)) {
+		eprintf ("failed to set PC!\n");
+		return false;
+	}
+	if (!r_debug_reg_sync (dbg, R_REG_TYPE_GPR, true)) {
+		eprintf ("cannot set registers!\n");
+		return false;
+	}
+
+	/* if we are on a software stepping breakpoint, we hide what is going on... */
+	if (b->swstep) {
+		dbg->reason.bp_addr = 0;
+		return true;
+	}
+
+	/* setup our stage 2 */
+	dbg->reason.bp_addr = b->addr;
+
+	/* check if cur bp demands tracing or not */
+	if (b->trace)
+		eprintf("hit tracepoint at: %" PFMT64x "\n", pc);
+	else
+		eprintf("hit breakpoint at: %" PFMT64x "\n", pc);
+
+	/* now that we've cleaned up after the breakpoint, call the other
+	 * potential breakpoint handlers
+	 */
+	if (dbg->corebind.core && dbg->corebind.bphit) {
+		dbg->corebind.bphit (dbg->corebind.core, b);
+	}
+
+	/* XXX(jjd): i don't think this goes here...
+	if (b->trace) {
+		r_debug_step (dbg, 1);
+		goto repeat;
+	}
+	 */
+	return true;
+}
+
+/*
+ * replace breakpoints before we continue execution
+ *
+ * this is called from r_debug_step_hard or r_debug_continue_kill
+ *
+ * this is a trick process because of breakpoints/tracepoints.
+ *
+ * if a breakpoint was just hit, we need step over that instruction before
+ * allowing the caller to proceed as desired.
+ *
+ * if the user wants to step, the single step here does the job.
+ */
+static int r_debug_recoil(RDebug *dbg, RDebugRecoilMode rc_mode) {
+	/* if bp_addr is not set, we must not have actually hit a breakpoint */
+	if (!dbg->reason.bp_addr) {
+		goto enable_breakpoints;
```

Comment:
uhm i dont like this goto much here. maybe just return enable_breakpoints(); and put that code in a separate static function